src/caffe/util/gpu_memory.cpp

```diff
@@ -37,7 +37,7 @@ void GPUMemory::Manager::init(const vector<int>& gpus, Mode m, bool debug) {
       // Just in case someone installed 'no cleanup' arena before
       delete cub_allocator_;
       cub_allocator_ = new cub::CachingDeviceAllocator(BIN_GROWTH, MIN_BIN,
-          MAX_BIN, MAX_CACHED_BYTES, false, debug_);
+          MAX_BIN, MAX_CACHED_BYTES, true, debug_);
```

Comment:
Right, so not having cleanup in the allocator destructor is fine. And yes we'd still run into the timing issues with other objects being destructed on exit and trying to return GPU memory. 
We can in fact take it a bit further and first time you get this 'cudart being unloaded' error, set some flag in mgr so that we don't even try to issue any cuda calls from that time on.